src/solvers/solver.py

```diff
@@ -0,0 +1,107 @@
+        """Uses the selected method for finding the solution."""
+
+    def prepare_cost_function(self, binary_variables: list) -> Tuple[Any, List[Any], List[Any]]:
+        """Uses the QUBO definition of the Workflow Scheduling Problem
+        to create the cost function with constraints."""
+
+        tasks_amount = len(self.tasks)
+
+        # Problem definition
+        cost_function = None
+        for cost, variable in zip(np.array(self.cost_df).flatten(), binary_variables):
+            if cost_function:
+                cost_function = cost_function + cost * variable
+            else:
+                cost_function = cost * variable
+
+        # Has to use exactly one machine
+        constraint_one_machine = []
+        for i in range(tasks_amount):
+            one_machine = None
+            for j in range(i, len(binary_variables), tasks_amount):
+                if one_machine:
+                    one_machine = one_machine + binary_variables[j]
+                else:
+                    one_machine = binary_variables[j]
+            constraint_one_machine.append(one_machine)
+
+        # All paths have to finish before the dedline
+        flat_runtimes = [(runtime, name) for n, machine_runtimes in self.runtimes_df.items() for runtime, name in
+                         zip(machine_runtimes, [j.name for j in self.tasks])]
+
+        constraint_path_runtime = []
+        for path in self.paths:
+            path_runtime = None
+            for var, (runtime, name) in zip(binary_variables, flat_runtimes):
+                if name not in path:
+                    continue
+                if path_runtime:
+                    path_runtime = path_runtime + runtime * var
+                else:
+                    path_runtime = runtime * var
+            constraint_path_runtime.append(path_runtime)
+
+        return cost_function, constraint_one_machine, constraint_path_runtime
+
+    def save_result(self, solution: Dict[str, str]) -> None:
+        """Assigns the correct machine names for the tasks."""
+        output = copy.deepcopy(self.wfcommons_data)
+        final_tasks = []
+        for task in self.wfcommons_data['workflow']['tasks']:
+            final_task = task.copy()
+            final_task['machine'] = solution[final_task['name']]
+            final_tasks.append(final_task)
+        output['workflow']['tasks'] = final_tasks
+        output['workflow']['machines'] = self.machines
+
+        write_json_file(output, self.output_filename)
+
+
+def set_machines(wfcommons_filename: str, machines_filename: str) -> Dict[str, dict]:
+    """Gets the data related to the machines either from the separate file
+    containg machine's details or from the original WfCommons data."""
+    if machines_filename is not None:
+        machines = get_machines_from_wfcommons_file(machines_filename)
+    else:
+        machines = get_machines_from_wfcommons_file(wfcommons_filename)
+    return machines
```

Comment:
```suggestion
    return get_machines_from_wfcommons_file(machines_filename or wfcommons_filename)
```
That should work heh